pkg/controller/master/storagenetwork/storage_network.go

```diff
@@ -0,0 +1,845 @@
+	if oldNad == "" {
+		return nil
+	}
+
+	nadName := strings.Split(oldNad, "/")
+	if len(nadName) != 2 {
+		// ignore this error and skip
+		logrus.Warnf("split nad namespace and name failed")
+		setting.Annotations[OldNadStorageNetworkAnnotation] = ""
+		return nil
+	}
+	namespace := nadName[0]
+	name := nadName[1]
+
+	if _, err := h.networkAttachmentDefinitionsCache.Get(namespace, name); err != nil {
+		if apierrors.IsNotFound(err) {
+			setting.Annotations[OldNadStorageNetworkAnnotation] = ""
+			return nil
+		}
+
+		// retry again
+		logrus.Warnf("check net-attach-def existing error %v", err)
+		return err
+	}
+
+	if err := h.networkAttachmentDefinitions.Delete(namespace, name, &metav1.DeleteOptions{}); err != nil {
+		logrus.Warnf("remove nad error %v", err)
+		return err
+	}
+
+	setting.Annotations[OldNadStorageNetworkAnnotation] = ""
+	return nil
+}
+
+// return true as finished
+func (h *Handler) checkLonghornSetting(setting *harvesterv1.Setting) bool {
+	value, err := h.getLonghornStorageNetwork()
+	if err != nil {
+		logrus.Warnf("get Longhorn settings error: %v", err)
+		return false
+	}
+
+	currentNad := setting.Annotations[NadStorageNetworkAnnotation]
+	if currentNad == value {
+		// check if we need to restart monitoring pods
+		if !h.checkPodStatusAndStart() {
+			if _, err := h.setConfiguredCondition(setting, false, ReasonInProgress, MsgRestartPod); err != nil {
+				logrus.Warnf("update status error %v", err)
+			}
+			h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+			return true
+		}
+
+		settingCopy := setting.DeepCopy()
+		if err := h.removeOldNad(settingCopy); err != nil {
+			logrus.Warnf("remove old nad error %v", err)
+			h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+			return true
+		}
+
+		// finish
+		if _, err := h.setConfiguredCondition(settingCopy, true, ReasonCompleted, ""); err != nil {
+			logrus.Warnf("update status error %v", err)
+		}
+		return true
+	}
+	return false
+}
+
+// true: all detach
+func (h *Handler) checkLonghornVolumeDetach() (bool, error) {
+	volumes, err := h.longhornVolumeCache.List(util.LonghornSystemNamespaceName, labels.Everything())
+	if err != nil {
+		logrus.Warnf("volume error %v", err)
+		return false, err
+	}
+
+	for _, volume := range volumes {
+		logrus.Infof("volume state: %v", volume.Status.State)
+		if volume.Status.State != "detached" {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (h *Handler) checkPrometheusStatusAndStart() bool {
+	// check prometheus cattle-monitoring-system/rancher-monitoring-prometheus replica
+	prometheus, err := h.prometheusCache.Get(CattleMonitoringSystemNamespace, RancherMonitoringPrometheus)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("prometheus not found. skip")
+			return true
+		}
+		logrus.Warnf("prometheus get error %v", err)
+		return false
+	}
+
+	logrus.Infof("prometheus: %v", *prometheus.Spec.Replicas)
+	// check started or not
+	if *prometheus.Spec.Replicas == 0 {
```

Comment:
We should disallow user config prometheus manually.
I will change this logic to another.

Reply:
It seems difficult to forbid the user to use `kubeclt` to set replica before, namely, before setting storage network, the user has already stopped `prometheus`.

Because you add `prometheus.Annotations[ReplicaStorageNetworkAnnotation]` to record the replica when `stop`, the `start` should stick to this annotation. Skip `start` when it's not there.

Reply:
I change the method.
please check it again thanks a lot